LibraBFT/Concrete/Properties/VotesOnce.agda

```diff
@@ -106,274 +106,99 @@ module LibraBFT.Concrete.Properties.VotesOnce where
-      → let pool = msgPool (StepPeer-post pstep)
-         in Meta-Honest-PK pk
-          → (ver  : WithVerSig pk v )(sv  : MsgWithSig∈ pk (ver-signature ver ) pool)
-          → (msgSender sv  , msgWhole sv)  ∈ List-map (pid ,_) outs
-          → (ver' : WithVerSig pk v')(sv' : MsgWithSig∈ pk (ver-signature ver') pool)
-          → (msgSender sv' , msgWhole sv') ∈ List-map (pid ,_) outs
-          → v ^∙ vEpoch ≡ v' ^∙ vEpoch
-          → v ^∙ vRound ≡ v' ^∙ vRound
-          → v ^∙ vProposedId ≡ v' ^∙ vProposedId
-    PredStep-hh preach (step-cheat fm isCheat) hip hpk ver sv (here refl) ver' sv' (here refl) epoch≡ r≡
-      with isCheat (msg⊆ sv) (msgSigned sv)
-    ...| inj₁ abs    = ⊥-elim (hpk abs) -- The key was honest by hypothesis.
-    ...| inj₂ sentb4
-      with isCheat (msg⊆ sv') (msgSigned sv')
-    ...| inj₁ abs     = ⊥-elim (hpk abs) -- The key was honest by hypothesis.
-    ...| inj₂ sentb4'
-      with msgSameSig sv | msgSameSig sv'
-    ...| refl | refl = hip hpk ver sentb4 ver' sentb4' epoch≡ r≡
-    PredStep-hh preach (step-honest x) hip hpk ver sv thisStep ver' sv' thisStep' epoch≡ r≡
-      with sameHonestSig⇒sameVoteData hpk ver (msgSigned sv) (sym (msgSameSig sv))
-    ...| inj₁ abs  = ⊥-elim (meta-sha256-cr abs)
+    -- Case (i) is trivial; cases (iii) and (iv) are symmetric and reduce to an implementation
+    -- obligation (Impl-VO1) and case (ii) reduces to a different implementation obligation
+    -- (Impl-VO2).
+
+
+    VotesOnceProof :
+       ∀ {v v' e pk} {st : SystemState e}
+       → ReachableSystemState st
+       → Meta-Honest-PK pk
+       → (vv  : WithVerSig pk v)  → MsgWithSig∈ pk (ver-signature vv) (msgPool st)
+       → (vv' : WithVerSig pk v') → MsgWithSig∈ pk (ver-signature vv') (msgPool st)
+       → v ^∙ vEpoch ≡ v' ^∙ vEpoch
+       → v ^∙ vRound ≡ v' ^∙ vRound
+       → v ^∙ vProposedId ≡ v' ^∙ vProposedId
+    VotesOnceProof step-0 _ _ msv _ _ _ _ = ⊥-elim (¬Any[] (msg∈pool msv))
+    VotesOnceProof (step-s r (step-epoch _)) pkH vv msv vv' msv' ep≡ r≡
+      = VotesOnceProof r pkH vv msv vv' msv' ep≡ r≡
+    VotesOnceProof (step-s r (step-peer cheat@(step-cheat f c))) pkH vv msv vv' msv' ep≡ r≡
+      with ¬cheatForgeNew cheat refl unit pkH msv | ¬cheatForgeNew cheat refl unit pkH msv'
+    ...| msb4 | m'sb4
+      with  msgSameSig msb4 | msgSameSig m'sb4
+    ...| refl | refl = VotesOnceProof r pkH vv msb4 vv' m'sb4 ep≡ r≡
+    VotesOnceProof (step-s r (step-peer (step-honest stPeer))) pkH vv msv vv' msv' ep≡ r≡
+        with  msgSameSig msv | msgSameSig msv'
+    ...| refl       | refl
+       with sameHonestSig⇒sameVoteData pkH (msgSigned msv) vv (msgSameSig msv)
+          | sameHonestSig⇒sameVoteData pkH (msgSigned msv') vv' (msgSameSig msv')
+    ...| inj₁ hb    | _         = ⊥-elim (meta-sha256-cr hb)
+    ...| inj₂ refl  | inj₁ hb   = ⊥-elim (meta-sha256-cr hb)
+    ...| inj₂ refl  | inj₂ refl
+      with newMsg⊎msgSent4 r stPeer pkH (msgSigned msv) (msg⊆ msv) (msg∈pool msv)
```

Comment:
I assume you intended to call this `newMsg⊎msgSentB4`?

Reply:
You are right!